### DISCUSSION TITLE
Tracking changes of devicePixelRatio

### DISCUSSION THREAD

**Discussion by heguro** [konvajs/konva#1709] [Ideas]
Currently, Konva sets the pixelRatio (canvas scaling factor) by [referencing](https://github.com/konvajs/konva/blob/a1660e1ccbd18c8b1b7d7303ea25f7a14a2d7858/src/Global.ts#L105) `window.devicePixelRatio` at the initial rendering by default, but does not adjust to subsequent changes in devicePixelRatio.  
Therefore, the display of the konva stage becomes blurred when the browser's zoom level changes or when the browser window is moved to a display with a different DPI.  
In the project I am involved in, we address this issue by redoing the setPixelRatio each time the devicePixelRatio changes.

examples: [CodePen](https://codepen.io/heguro/pen/OJqXWqJ) / [CodeSandbox (with React)](https://codesandbox.io/p/sandbox/konva-pixelratio-reset-react-59pjw8)

Is it possible for this behavior to be implemented internally in Konva?

may related: #1620

**Comment by lavrton** [comment]
It may be a good idea to implement that. However, it may be a breaking change, and we should do it carefully because some users may prefer to manual control the pixel ratio of a layer.

Personally, I am not interested in the functionality right now, but I am very open to good Pull Requests.

**Comment by psychedelicious** [comment]
I just ran into this in my react app, and implemented something similar to @heguro's example.

In an effort to simplify my code, I found this hook that encapsulates the logic and provides some additional functionality: https://github.com/rexxars/use-device-pixel-ratio

```ts
const dpr = useDevicePixelRatio({ round: false /* maybe required for perfect scaling? */ });
useLayoutEffect(() => {
  Konva.pixelRatio = dpr;
}, [dpr]);
```

All konva stages in the app get the correct pixel ratio whenever the user changes page zoom.
